Bug: IFBO

EDIT: Not windows specific, will look into it

* See: https://github.com/automl/neps/actions/runs/12976865310/job/36189616752

Seems to be two test failures from #182 which both occur on windows and both for an out-of-bounds index when taking back a number from ifbo and turning it into the specific categorical value at that index.

Might be a bug on my part, however the tests in #182 will detect if it happens again on other platforms or it's windows specific.

```python
optimizer_and_key_and_search_space = (IFBO(space=SearchSpace(elements={'a': Float(lower=0.0, upper=1.0, center=0.5, log=False, prior=0.5, prior_confidence=...(lower=1, upper=10, center=6, log=False, prior=None, prior_confidence='low', is_fidelity=True)}, constants={'c': 'a'}))

    def test_optimizers_work_roughly(
        optimizer_and_key_and_search_space: tuple[AskFunction, str, SearchSpace],
    ) -> None:
        opt, key, search_space = optimizer_and_key_and_search_space
        ask_and_tell = AskAndTell(opt)
    
        for _ in range(20):
>           trial = ask_and_tell.ask()

tests\test_state\test_neps_state.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
neps\optimizers\ask_and_tell.py:133: in ask
    sampled_config = self.optimizer(self.trials, budget_info, n)
neps\optimizers\ifbo.py:241: in __call__
    _id, fid, config = decode_ftpfn_data(
neps\optimizers\models\ftpfn.py:210: in decode_ftpfn_data
    configs = encoder.decode(x[:, 2:])
neps\space\encoding.py:416: in decode
    values[hp_name] = transformer.decode(tensor)
neps\space\encoding.py:157: in decode
    return self._integer_transformer.decode(x)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = CategoricalToIntegerTransformer(choices=['a', 'b', 'c'], domain=Domain(lower=0, upper=2, round=True, log_bounds=None, bins=None, is_categorical=True), _lookup=None)
x = tensor([4])

    @override
    def decode(self, x: torch.Tensor) -> list[Any]:
>       return [self.choices[int(i)] for i in torch.round(x).tolist()]
E       IndexError: list index out of range

neps\space\encoding.py:108: IndexError
```

--- a/neps/optimizers/bayesian_optimization.py
+++ b/neps/optimizers/bayesian_optimization.py
@@ -1,5 +1,6 @@
 from __future__ import annotations
 
+import itertools
 import math
 from collections.abc import Mapping
 from dataclasses import dataclass
@@ -89,9 +90,7 @@ class BayesianOptimization:
 
         n_to_sample = 1 if n is None else n
         n_sampled = len(trials)
-        config_ids = iter(
-            str(i + 1) for i in range(n_sampled, n_sampled + n_to_sample + 1)
-        )
+        id_generator = iter(str(i) for i in itertools.count(n_sampled + 1))
 
         # If the amount of configs evaluated is less than the initial design
         # requirement, keep drawing from initial design
@@ -103,6 +102,8 @@ class BayesianOptimization:
         sampled_configs: list[SampledConfig] = []
 
         if n_evaluated < self.n_initial_design:
+            # For reproducibility, we need to ensure we do the same sample of all
+            # configs each time.
             design_samples = make_initial_design(
                 parameters=parameters,
                 encoder=self.encoder,
@@ -111,13 +112,21 @@ class BayesianOptimization:
                 seed=None,  # TODO: Seeding, however we need to avoid repeating configs
                 sample_size=self.n_initial_design,
             )
+
+            # Then take the subset we actually need
+            design_samples = design_samples[n_evaluated:]
             for sample in design_samples:
                 sample.update(self.space.constants)
 
             sampled_configs.extend(
                 [
                     SampledConfig(id=config_id, config=config)
-                    for config_id, config in zip(config_ids, design_samples, strict=True)
+                    for config_id, config in zip(
+                        id_generator,
+                        design_samples,
+                        # NOTE: We use a generator for the ids so no need for strict
+                        strict=False,
+                    )
                 ]
             )
 
@@ -187,7 +196,12 @@ class BayesianOptimization:
         sampled_configs.extend(
             [
                 SampledConfig(id=config_id, config=config)
-                for config_id, config in zip(config_ids, configs, strict=True)
+                for config_id, config in zip(
+                    id_generator,
+                    configs,
+                    # NOTE: We use a generator for the ids so no need for strict
+                    strict=False,
+                )
             ]
         )
         return sampled_configs[0] if n is None else sampled_configs

--- a/neps/optimizers/bayesian_optimization.py
+++ b/neps/optimizers/bayesian_optimization.py
@@ -89,7 +89,9 @@ class BayesianOptimization:
 
         n_to_sample = 1 if n is None else n
         n_sampled = len(trials)
-        config_ids = iter(str(i + 1) for i in range(n_sampled, n_sampled + n_to_sample))
+        config_ids = iter(
+            str(i + 1) for i in range(n_sampled, n_sampled + n_to_sample + 1)
+        )
 
         # If the amount of configs evaluated is less than the initial design
         # requirement, keep drawing from initial design

--- a/neps/optimizers/bayesian_optimization.py
+++ b/neps/optimizers/bayesian_optimization.py
@@ -98,6 +98,8 @@ class BayesianOptimization:
             for trial in trials.values()
             if trial.report is not None and trial.report.objective_to_minimize is not None
         )
+        sampled_configs: list[SampledConfig] = []
+
         if n_evaluated < self.n_initial_design:
             design_samples = make_initial_design(
                 parameters=parameters,
@@ -105,16 +107,20 @@ class BayesianOptimization:
                 sample_prior_first=self.sample_prior_first if n_sampled == 0 else False,
                 sampler=self.prior if self.prior is not None else "uniform",
                 seed=None,  # TODO: Seeding, however we need to avoid repeating configs
-                sample_size=n_to_sample,
+                sample_size=self.n_initial_design,
             )
             for sample in design_samples:
                 sample.update(self.space.constants)
 
-            sampled_configs = [
-                SampledConfig(id=config_id, config=config)
-                for config_id, config in zip(config_ids, design_samples, strict=True)
-            ]
-            return sampled_configs[0] if n is None else sampled_configs
+            sampled_configs.extend(
+                [
+                    SampledConfig(id=config_id, config=config)
+                    for config_id, config in zip(config_ids, design_samples, strict=True)
+                ]
+            )
+
+            if len(sampled_configs) >= n_to_sample:
+                return sampled_configs[0] if n is None else sampled_configs
 
         # Otherwise, we encode trials and setup to fit and acquire from a GP
         data, encoder = encode_trials_for_gp(
@@ -149,6 +155,7 @@ class BayesianOptimization:
             prior = None if pibo_exp_term < 1e-4 else self.prior
 
         gp = make_default_single_obj_gp(x=data.x, y=data.y, encoder=encoder)
+        n_to_acquire = n_to_sample - len(sampled_configs)
         candidates = fit_and_acquire_from_gp(
             gp=gp,
             x_train=data.x,
@@ -164,7 +171,7 @@ class BayesianOptimization:
                 prune_baseline=True,
             ),
             prior=prior,
-            n_candidates_required=n_to_sample,
+            n_candidates_required=n_to_acquire,
             pibo_exp_term=pibo_exp_term,
             costs=data.cost if self.cost_aware is not False else None,
             cost_percentage_used=cost_percent,
@@ -175,8 +182,10 @@ class BayesianOptimization:
         for config in configs:
             config.update(self.space.constants)
 
-        sampled_configs = [
-            SampledConfig(id=config_id, config=config)
-            for config_id, config in zip(config_ids, configs, strict=True)
-        ]
+        sampled_configs.extend(
+            [
+                SampledConfig(id=config_id, config=config)
+                for config_id, config in zip(config_ids, configs, strict=True)
+            ]
+        )
         return sampled_configs[0] if n is None else sampled_configs

--- a/neps/optimizers/ifbo.py
+++ b/neps/optimizers/ifbo.py
@@ -140,7 +140,7 @@ class IFBO:
 
         assert n is None, "TODO"
         ids = [int(config_id.split("_", maxsplit=1)[0]) for config_id in trials]
-        new_id = max(ids) + 1 if len(ids) > 0 else 0
+        new_id = max(ids) + 1 if len(ids) > 0 else 1
 
         # The FTPFN surrogate takes in a budget in the range [0, 1]
         # We also need to be able to map these to discrete integers
@@ -153,7 +153,7 @@ class IFBO:
         budget_index_domain = Domain.indices(self.n_fidelity_bins)
 
         # If we havn't passed the intial design phase
-        if new_id < self.n_initial_design:
+        if new_id <= self.n_initial_design:
             init_design = make_initial_design(
                 parameters=parameters,
                 encoder=self.encoder,
@@ -163,7 +163,7 @@ class IFBO:
                 sample_size=self.n_initial_design,
             )
 
-            config = init_design[new_id]
+            config = init_design[new_id - 1]
             config[fidelity_name] = fidelity.lower
             config.update(self.space.constants)
             return SampledConfig(id=f"{new_id}_0", config=config)
@@ -238,13 +238,16 @@ class IFBO:
             local_search_sample_size=256,
             local_search_confidence=0.95,
         )
+
         _id, fid, config = decode_ftpfn_data(
             best_row,
-            self.encoder,
+            encoder=self.encoder,
             budget_domain=budget_domain,
             fidelity_domain=fidelity.domain,
         )[0]
 
+        # If _id is None, that means a new config was sampled and
+        # should be evaluated one step
         if _id is None:
             config[fidelity_name] = fid
             config.update(self.space.constants)

--- a/neps/optimizers/ifbo.py
+++ b/neps/optimizers/ifbo.py
@@ -23,6 +23,7 @@ if TYPE_CHECKING:
 
 # NOTE: Ifbo was trained using 32 bit
 FTPFN_DTYPE = torch.float32
+BUDGET_DOMAIN_EPS = 1e-6
 
 
 def _adjust_space_to_match_stepsize(
@@ -147,10 +148,24 @@ class IFBO:
         # Hence we use the two domains below to do so.
 
         # Domain in which we should pass budgets to ifbo model
-        budget_domain = Domain.floating(lower=1 / fidelity.upper, upper=1)
+        # IFBO expects them to be in `(0 1]`, where explicitly 0 is
+        # not allowed. We map this close to `BUDGET_DOMAIN_EPS` depending
+        # on the scale of the fidelity domain.
+        budget_domain = Domain.floating(
+            lower=(fidelity.lower + BUDGET_DOMAIN_EPS)
+            / (fidelity.upper - fidelity.lower),
+            upper=1,
+        )
+
+        # However, we need to make sure we don't end up with a positive
+        # `lower=` which gauranteed with this assertion.
+        if fidelity.upper - fidelity.lower < BUDGET_DOMAIN_EPS:
+            raise ValueError(
+                f"Fidelity domain {fidelity} is too small to be used with ifBO."
+            )
 
         # Domain from which we assign an index to each budget
-        budget_index_domain = Domain.indices(self.n_fidelity_bins)
+        budget_index_domain = Domain.indices(self.n_fidelity_bins + 1)
 
         # If we havn't passed the intial design phase
         if new_id <= self.n_initial_design:

--- a/neps/runtime.py
+++ b/neps/runtime.py
@@ -464,30 +464,29 @@ class DefaultWorker(Generic[Loc]):
                     return this_workers_trial
                 except TrialAlreadyExistsError as e:
                     if e.trial_id in trials:
-                        logger.error(
-                            "The new sampled trial was given an id of '%s', yet this"
-                            " exists in the loaded in trials given to the optimizer. This"
-                            " indicates a bug with the optimizers allocation of ids.",
-                            e.trial_id,
-                        )
-                    else:
-                        _grace = DefaultWorker._GRACE
-                        _inc = FS_SYNC_GRACE_INC
-                        logger.warning(
-                            "The new sampled trial was given an id of '%s', which is not"
-                            " one that was loaded in by the optimizer. This is usually"
-                            " an indication that the file-system you are running on"
-                            " is not atmoic in synchoronizing file operations."
-                            " We have attempted to stabalize this but milage may vary."
-                            " We are incrementing a grace period for file-locks from"
-                            " '%s's to '%s's. You can control the initial"
-                            " grace with 'NEPS_FS_SYNC_GRACE_BASE' and the increment with"
-                            " 'NEPS_FS_SYNC_GRACE_INC'.",
-                            e.trial_id,
-                            _grace,
-                            _grace + _inc,
-                        )
-                        DefaultWorker._GRACE = _grace + FS_SYNC_GRACE_INC
+                        raise RuntimeError(
+                            f"The new sampled trial was given an id of {e.trial_id}, yet"
+                            " this exists in the loaded in trials given to the optimizer."
+                            " This is a bug with the optimizers allocation of ids."
+                        ) from e
+
+                    _grace = DefaultWorker._GRACE
+                    _inc = FS_SYNC_GRACE_INC
+                    logger.warning(
+                        "The new sampled trial was given an id of '%s', which is not"
+                        " one that was loaded in by the optimizer. This is usually"
+                        " an indication that the file-system you are running on"
+                        " is not atmoic in synchoronizing file operations."
+                        " We have attempted to stabalize this but milage may vary."
+                        " We are incrementing a grace period for file-locks from"
+                        " '%s's to '%s's. You can control the initial"
+                        " grace with 'NEPS_FS_SYNC_GRACE_BASE' and the increment with"
+                        " 'NEPS_FS_SYNC_GRACE_INC'.",
+                        e.trial_id,
+                        _grace,
+                        _grace + _inc,
+                    )
+                    DefaultWorker._GRACE = _grace + FS_SYNC_GRACE_INC
                     raise e
 
     # Forgive me lord, for I have sinned, this function is atrocious but complicated

--- a/neps/space/encoding.py
+++ b/neps/space/encoding.py
@@ -72,11 +72,10 @@ class CategoricalToIntegerTransformer(TensorTransformer):
 
         self.domain = Domain.indices(len(self.choices), is_categorical=True)
         self._lookup = None
-        if len(self.choices) > 3:
-            try:
-                self._lookup = {c: i for i, c in enumerate(self.choices)}
-            except TypeError:
-                self._lookup = None
+        try:
+            self._lookup = {c: i for i, c in enumerate(self.choices)}
+        except TypeError:
+            self._lookup = None
 
     @override
     def encode(
@@ -117,8 +116,9 @@ class CategoricalToUnitNorm(TensorTransformer):
 
     choices: Sequence[Any]
 
-    domain: Domain = field(init=False)
-    _integer_transformer: CategoricalToIntegerTransformer = field(init=False)
+    domain: Domain[float] = field(init=False)
+    _cat_int_domain: Domain[int] = field(init=False)
+    _lookup: dict[Any, int] | None = field(init=False)
 
     def __post_init__(self) -> None:
         self.domain = Domain.floating(
@@ -127,7 +127,11 @@ class CategoricalToUnitNorm(TensorTransformer):
             bins=len(self.choices),
             is_categorical=True,
         )
-        self._integer_transformer = CategoricalToIntegerTransformer(self.choices)
+        self._cat_int_domain = Domain.indices(len(self.choices), is_categorical=True)
+        try:
+            self._lookup = {c: i for i, c in enumerate(self.choices)}
+        except TypeError:
+            self._lookup = None
 
     @override
     def encode(
@@ -138,13 +142,19 @@ class CategoricalToUnitNorm(TensorTransformer):
         dtype: torch.dtype | None = None,
         device: torch.device | None = None,
     ) -> torch.Tensor:
-        integers = self._integer_transformer.encode(
-            x,
-            dtype=dtype if dtype is not None else torch.float64,
-            device=device,
+        if dtype is None:
+            dtype = torch.int if out is None else out.dtype
+
+        values = (
+            [self._lookup[c] for c in x]
+            if self._lookup
+            else [self.choices.index(c) for c in x]
         )
+        integers = torch.tensor(values, dtype=torch.int64, device=device)
         binned_floats = self.domain.cast(
-            integers, frm=self._integer_transformer.domain, dtype=dtype
+            integers,
+            frm=self._cat_int_domain,
+            dtype=dtype,
         )
         if out is not None:
             return out.copy_(binned_floats)
@@ -153,8 +163,8 @@ class CategoricalToUnitNorm(TensorTransformer):
 
     @override
     def decode(self, x: torch.Tensor) -> list[Any]:
-        x = torch.round(x * (len(self.choices) - 1)).type(torch.int64)
-        return self._integer_transformer.decode(x)
+        x = self._cat_int_domain.cast(x, frm=self.domain)
+        return [self.choices[int(i)] for i in torch.round(x).tolist()]
 
 
 # TODO: Maybe add a shift argument, could be useful to have `0` as midpoint

--- a/neps/space/parameters.py
+++ b/neps/space/parameters.py
@@ -75,6 +75,13 @@ class Float:
         self.lower = float(self.lower)
         self.upper = float(self.upper)
 
+        if self.is_fidelity and (self.lower < 0 or self.upper < 0):
+            raise ValueError(
+                f"Float parameter: fidelity bounds error. Expected fidelity"
+                f" bounds to be >= 0, but got lower={self.lower}, "
+                f" upper={self.upper}."
+            )
+
         if np.isnan(self.lower):
             raise ValueError("Can not have lower bound that is nan")
 
@@ -148,6 +155,13 @@ class Integer:
 
         self.lower = lower_int
         self.upper = upper_int
+
+        if self.is_fidelity and (self.lower < 0 or self.upper < 0):
+            raise ValueError(
+                f"Integer parameter: fidelity bounds error. Expected fidelity"
+                f" bounds to be >= 0, but got lower={self.lower}, "
+                f" upper={self.upper}."
+            )
 
         if self.log and (self.lower <= 0 or self.upper <= 0):
             raise ValueError(

